core/src/main/java/org/apache/iceberg/rest/auth/AuthSession.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.auth;
+
+import java.util.Map;
+
+public interface AuthSession {
+
+  Map<String, String> headers();
```

Comment:
As I stated in a previous comment, this is where I think we need something more powerful than just returning some headers to include in the request. This is because some auth schemes will need to inspect the whole request and make decisions based on its content/state.

Here is what I have currently:

```java
  void applyAuth(
      URI requestUri,
      String method,
      Map<String, String> queryParams,
      Object requestBody,
      Map<String, List<String>> currentHeaders,
      BiConsumer<String, String> headerConsumer);
```

Most impls would simply inject additional headers:

```java
    headerConsumer.accept("Authorization", "Bearer ****");
```

Reply:
@amogh-jahagirdar you are spot on. I was asking myself the same question. 

There are pros and cons for each approach:

1. Higher level 
    1. pros: integrates well with the catalog sessions and allows for context-level or table-level credentials (as it exists today with OAuth).
    2. cons: cannot leverage the existing machinery around `CredentialsProvider`; could potentially interfere with proxy credentials which are implemented in lower level.
2. Lower level 
    1. pros: very easy to implement all auth schemes using either `CredentialProvider` or `HttpRequestInterceptor`. Interference with proxy credentials could be easily remediated.
    2. cons: cannot integrate with catalog sessions; context-level or table-level sessions _cannot_ use specialized credentials.

My personal take is that the cons of a lower level approach are a blocker: we cannot give up of what's already built around table-level credentials. So I'd suggest to pursue the higher level approach that Jack initiated here.